### PR TITLE
[Builder] Add mlrun requirement when installing python packages also on serving functions

### PR DESCRIPTION
**Pull request by roei3000b**
Continue to this pr: https://github.com/mlrun/mlrun/pull/4623
Add mlrun requirements also in serving function flow.

https://iguazio.atlassian.net/browse/ML-6987

### CODE REVIEW

File: server/api/crud/runtimes/nuclio/function.py [hunk 1]
```diff
@@ -36,7 +36,9 @@
 import server.api.utils.builder
 import server.api.utils.clients.async_nuclio
 import server.api.utils.clients.iguazio
+import server.api.utils.helpers
 import server.api.utils.singletons.k8s
+from mlrun.config import config as mlrun_config
```
Comment (alonmr) [inline server/api/crud/runtimes/nuclio/function.py]:
```suggestion
from mlrun import mlconf
```

File: server/api/utils/builder.py [hunk 2]
```diff
@@ -775,6 +755,36 @@ def build_runtime(
...
+
+def is_mlrun_image(base_image):
+    mlrun_images = [
+        "mlrun/mlrun",
+        "mlrun/mlrun-gpu",
+        "mlrun/ml-base",
+    ]
+    if any([image in base_image for image in mlrun_images]):
+        return True
+    return False
```
Comment (alonmr) [inline server/api/utils/builder.py]:
```suggestion
    return any([image in base_image for image in mlrun_images])
```

File: tests/api/runtimes/test_nuclio.py [hunk 3]
```diff
@@ -603,6 +603,21 @@ def test_deploy_with_security_context_enrichment(
             )
         )
 
+    def test_deploy_mlrun_requirements(
+        self, db: Session, k8s_secrets_mock: K8sSecretsMock
+    ):
+        user_unix_id = 1000
```
Comment (alonmr) [inline tests/api/runtimes/test_nuclio.py]: I'm not sure if `user_unix_id` and `igz_version` are needed. see if you can remove them

**Review by alonmr — COMMENTED** [review]
Well done! Some small comments

File: server/api/crud/runtimes/nuclio/function.py [hunk 4]
```diff
@@ -324,6 +326,16 @@ def _compile_function_config(
...
 
+    build: mlrun.model.ImageBuilder = function.spec.build
+    base_image: str = (
+        build.base_image or function.spec.image or mlrun_config.default_base_image
+    )
+
+    if server.api.utils.builder.is_mlrun_image(base_image) and build.requirements:
+        enriched_base_image = function.full_image_path(
+            base_image, client_version, client_python_version
+        )
```
Comment (alonmr) [inline server/api/crud/runtimes/nuclio/function.py]: this is duplicated with `_resolve_and_set_base_image` below. see if you can move `_resolve_and_set_build_requirements_and_commands` after `_resolve_and_set_base_image` and I would also move the added code into `_resolve_and_set_build_requirements_and_commands`

Comment (roei3000b) [reply server/api/crud/runtimes/nuclio/function.py]: Do you mean that this is the duplication?
```
base_image = (
        mlrun.utils.get_in(config, "spec.build.baseImage")
        or function.spec.image
        or function.spec.build.base_image
    )
```

Comment (alonmr) [reply server/api/crud/runtimes/nuclio/function.py]: yes and the enrichment below

**Review by alonmr — APPROVED** [review]
lgtm

File: server/api/crud/runtimes/nuclio/function.py [hunk 5]
```diff
@@ -257,6 +258,9 @@ def _compile_function_config(
     builder_env=None,
     auth_info=None,
 ):
+    """
+    Compile the nuclio function configuration from the mlrun function object.
+    """
```
Comment (TomerShor) [inline server/api/crud/runtimes/nuclio/function.py]: Add the full docstring format, including params and returns

File: server/api/crud/runtimes/nuclio/function.py [hunk 6]
```diff
@@ -378,6 +381,17 @@ def _compile_function_config(
...
     _resolve_and_set_base_image(function, config, client_version, client_python_version)
+    build: mlrun.model.ImageBuilder = function.spec.build
+    base_image = mlrun.utils.get_in(config, "spec.build.baseImage")
+
+    if (
+        base_image
+        and server.api.utils.builder.is_mlrun_image(base_image)
+        and build.requirements
+    ):
+        server.api.utils.builder.add_mlrun_to_requirements(build, base_image)
```
Comment (TomerShor) [inline server/api/crud/runtimes/nuclio/function.py]: This section can also be moved inside `_resolve_and_set_build_requirements_and_commands` as it is also setting build requirements (+ it keeps the code even cleaner)

**Review by TomerShor — CHANGES_REQUESTED** [review]
Minor improvements

File: server/api/crud/runtimes/nuclio/function.py [hunk 7]
```diff
@@ -257,6 +258,17 @@ def _compile_function_config(
     builder_env=None,
     auth_info=None,
 ):
+    """
+    Compile the nuclio function configuration from the mlrun function object.
+
+    :param function:              nuclio function object
```
Comment (TomerShor) [inline server/api/crud/runtimes/nuclio/function.py]:
```suggestion
    :param function:              mlrun function object
```

File: server/api/crud/runtimes/nuclio/function.py [hunk 8]
```diff
@@ -257,6 +258,17 @@ def _compile_function_config(
...
+    :param client_version:        mlrun client version
+    :param client_python_version: mlrun client python version
+    :param builder_env:           mlrun builder environment (for config/credentials)
+    :param auth_info:             service AuthInfo
+
+    :return: function name, project name, function config
```
Comment (TomerShor) [inline server/api/crud/runtimes/nuclio/function.py]:
```suggestion
    :return: function name, project name, nuclio function config
```

**Review by TomerShor — CHANGES_REQUESTED** [review]

**Review by TomerShor — APPROVED** [review]

**Pull request merged**